flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/JobInitializationMetricsBuilder.java

```diff
@@ -19,40 +19,52 @@
 package org.apache.flink.runtime.checkpoint;
 
 import org.apache.flink.runtime.checkpoint.JobInitializationMetrics.SumMaxDuration;
+import org.apache.flink.runtime.executiongraph.Execution;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.ArrayList;
 import java.util.HashMap;
-import java.util.List;
+import java.util.HashSet;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
 
 import static org.apache.flink.runtime.checkpoint.JobInitializationMetrics.UNSET;
-import static org.apache.flink.util.Preconditions.checkArgument;
 import static org.apache.flink.util.Preconditions.checkState;
 
 class JobInitializationMetricsBuilder {
     private static final Logger LOG =
             LoggerFactory.getLogger(JobInitializationMetricsBuilder.class);
 
-    private final List<SubTaskInitializationMetrics> reportedMetrics = new ArrayList<>();
-    private final int totalNumberOfSubTasks;
+    private final Map<ExecutionAttemptID, SubTaskInitializationMetrics> reportedMetrics =
+            new HashMap<>();
+    private final Set<ExecutionAttemptID> toInitialize;
     private final long startTs;
     private Optional<Long> stateSize = Optional.empty();
     private Optional<Long> checkpointId = Optional.empty();
     private Optional<CheckpointProperties> checkpointProperties = Optional.empty();
     private Optional<String> externalPath = Optional.empty();
 
-    JobInitializationMetricsBuilder(int totalNumberOfSubTasks, long startTs) {
-        checkArgument(totalNumberOfSubTasks > 0);
-        this.totalNumberOfSubTasks = totalNumberOfSubTasks;
+    /**
```

Comment:
> I'm just unsure why we can replace the totalNumberOfSubTasks with a number of attempts. They seem like unrelated numbers to me. Can you please elaborate?

You have a fair point. I verified once more that we're actually doing the right thing. The change is fine. I added JavaDoc to help understanding the code here. Additionally, I added a description to the commit message which I will preserve. Please have a look whether my reasoning makes sense.

Reply:
Thanks for the verification.